src/cargo/core/gc.rs

```diff
@@ -0,0 +1,557 @@
+            self.max_src_age = Some(maybe_newer_span(max_download_age, self.max_src_age));
+            self.max_crate_age = Some(maybe_newer_span(max_download_age, self.max_crate_age));
+            self.max_index_age = Some(maybe_newer_span(max_download_age, self.max_index_age));
+            self.max_git_co_age = Some(maybe_newer_span(max_download_age, self.max_git_co_age));
+            self.max_git_db_age = Some(maybe_newer_span(max_download_age, self.max_git_db_age));
+        }
+        Ok(())
+    }
+}
+
+/// The kind of automatic garbage collection to perform.
+///
+/// "Automatic" is the kind of gc performed automatically by Cargo in any
+/// command that is already doing a bunch of work. See [`auto_gc`] for more.
+#[derive(Clone, Debug)]
+pub enum AutoGcKind {
+    /// Automatically clean up the downloaded files *and* the target directory.
+    ///
+    /// This is the mode used by default.
+    All,
+    /// Automatically clean only downloaded files.
+    ///
+    /// This corresponds to `cargo clean --gc=download`.
+    Download,
+    /// Automatically clean only the target directory.
+    ///
+    /// THIS IS NOT IMPLEMENTED.
+    ///
+    /// This corresponds to `cargo clean --gc=target`.
+    Target,
+    /// Automatically clean only the shared target directory.
+    ///
+    /// THIS IS NOT IMPLEMENTED.
+    ///
+    /// This corresponds to `cargo clean --gc=shared-target`.
+    SharedTarget,
+}
+
+/// Garbage collector.
+///
+/// See the module docs at [`crate::core::gc`] for more information on GC.
+pub struct Gc<'a, 'config> {
+    config: &'config Config,
+    global_cache_tracker: &'a mut GlobalCacheTracker,
+    /// A lock on the package cache.
+    ///
+    /// This is important to be held, since we don't want multiple cargos to
+    /// be allowed to write to the cache at the same time, or for others to
+    /// read while we are modifying the cache.
+    #[allow(dead_code)] // Held for drop.
+    lock: CacheLock<'config>,
+}
+
+impl<'a, 'config> Gc<'a, 'config> {
+    pub fn new(
+        config: &'config Config,
+        global_cache_tracker: &'a mut GlobalCacheTracker,
+    ) -> CargoResult<Gc<'a, 'config>> {
+        let lock = config.acquire_package_cache_lock(CacheLockMode::MutateExclusive)?;
+        Ok(Gc {
+            config,
+            global_cache_tracker,
+            lock,
+        })
+    }
+
+    /// Performs automatic garbage cleaning.
+    ///
+    /// This returns immediately without doing work if garbage collection has
+    /// been performed recently (since `gc.auto.frequency`).
+    fn auto(&mut self, clean_ctx: &mut CleanContext<'config>) -> CargoResult<()> {
+        if !self.config.cli_unstable().gc {
+            return Ok(());
+        }
+        let auto_config = self
+            .config
+            .get::<Option<AutoConfig>>("gc.auto")?
+            .unwrap_or_default();
+        let Some(freq) = parse_frequency(auto_config.frequency.as_deref().unwrap_or("1 day"))?
```

Comment:
nit: I feel like this constant should be with the definition of `AutoConfig` or pulled out into a separate place with the others

Reply:
I moved it to a module-level constant.

I'm not particularly happy with the way the defaults are defined, but I don't know of a better way. It could use serde defaults like:

```rust
struct AutoConfig {
    /// The maximum frequency that automatic garbage collection happens.
    #[serde(default="default_frequency")]
    frequency: String,
    #[serde(default="default_max_src_age")]
    max_src_age: String,
    // etc....
}

fn default_frequency() -> String {
    "1 day".to_string()
}

fn default_max_src_age() -> String {
    "1 month".to_string()
}
```

but that seems worse somehow (lots of lines of code, lots of unnecessary string allocations, though that is minor). Unless I'm missing something about serde.

Or, it could have accessors like:

```rust
fn frequency(&self) -> &str {
    self.frequency.as_deref().unwrap_or("1 day")
}
```

But again, that adds 20-30 lines of code.  I could do that, but since all fields are accessed in exactly one place, I'm not sure if it helps.

I could make those changes if you want, or if you have other ideas on how to deal with config defaults.

Reply:
For what this is doing, the `const`s are fine.